draft-ietf-tls-tls13.md

```diff
@@ -1953,11 +1947,14 @@ The "extension_data" field of this extension contains a
 %%% Key Exchange Messages
 
        struct {
-           select (role) {
-               case client:
+           select (Handshake.msg_type) {
+               case client_hello:
                    KeyShareEntry client_shares<0..2^16-1>;
 
-               case server:
+               case hello_retry_request:
+                   NamedGroup selected_group;
+
+               case server_hello:
                    KeyShareEntry server_share;
            }
        } KeyShare;
```

Comment:
> For BoringSSL we've intentionally had the server just pick a group without paying attention to the key share list. Then we'll decide whether we HelloRetryRequest based on that decision. It's possible, in the future, there will be more variation in curve preferences but, right now, the only sensible curves are X25519 and P-256 anyway.

We've done the same in NSS, though in future we will pick the curve and permit "equal" curves to be picked if the first choice doesn't have a key share.

I generally agree on the position that the encoding doesn't need to be the same for client and server (unless it needs to be).  Extensions aren't fixed structures, they are containers for a protocol.